### PR TITLE
fix sidecar sshd failed

### PR DESCRIPTION
**Pull request by shy-Xu**
**What this PR does / why we need it**:

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by shy-Xu** [comment]
fix issue #1377

File: internal/nhctl/controller/service.go [hunk 1]
```diff
@@ -268,6 +269,14 @@ func (c *Controller) getGeneratedDeploymentLabels() map[string]string {
 	}
 }
 
+func (c *Controller) sidecarContainerPrivileged(ops *model.DevStartOptions) bool {
+	if strings.Contains(ops.DevImage, "php") {
```
Comment (yuyicai) [inline internal/nhctl/controller/service.go:273]: Options has a variable: language.
You can use this variable to make a decision.

Comment (shy-Xu) [reply internal/nhctl/controller/service.go:273]: I can not find this variable in 'model.DevStartOptions'

File: internal/nhctl/controller/dev_start.go [hunk 2]
```diff
@@ -375,6 +376,12 @@ func generateSideCarContainer(sidecarImage, workDir string) corev1.Container {
 		WorkingDir: workDir,
 	}
 
+	if sshUsed {
+		var rootUID int64 = 0
+		sideCarContainer.SecurityContext = &corev1.SecurityContext{RunAsUser: &rootUID}
+		sideCarContainer.Image = _const.SSHSideCarImage
```
Comment (yuyicai) [inline internal/nhctl/controller/dev_start.go:382]: If user has customized the image, it should not be changed.

File: internal/nhctl/controller/service.go (hunk 1 above)
Comment (yuyicai) [reply internal/nhctl/controller/service.go:273]: > I can not find this variable in 'model.DevStartOptions'

https://github.com/nocalhost/nocalhost/blob/v0.6.18/internal/nhctl/profile/config_v2.go#L99
```
type DebugConfig struct {
	RemoteDebugPort int    `validate:"Port" json:"remoteDebugPort" yaml:"remoteDebugPort"`
	Language        string `validate:"Language" json:"language" yaml:"language"`
}
```

File: internal/nhctl/controller/dev_start.go [hunk 3]
```diff
@@ -603,7 +610,7 @@ func (c *Controller) genContainersAndVolumes(podSpec *corev1.PodSpec,
 		devImage = c.GetDevImage(containerName) // Default : replace the first container
 	}
 
-	sideCarContainer := generateSideCarContainer(c.GetDevSidecarImage(containerName), workDir)
+	sideCarContainer := generateSideCarContainer(c.GetDevSidecarImage(containerName), workDir, sshUsed)
```
Comment (yuyicai) [inline internal/nhctl/controller/dev_start.go:613]: You cat get all `ContainerDevConfig` from `*Controller`.  
No need to add `genContainersAndVolumes()` input parameter `sshUsed bool`

File: internal/nhctl/controller/dev_start.go [hunk 4]
```diff
@@ -561,7 +568,7 @@ func containerStatusForDevPod(pod *corev1.Pod, consumeFun func(status string, er
 }
 
 func (c *Controller) genContainersAndVolumes(podSpec *corev1.PodSpec,
-	containerName, devImage, storageClass string, duplicateDevMode bool) (*corev1.Container,
+	containerName, devImage, storageClass string, duplicateDevMode bool, sshUsed bool) (*corev1.Container,
```
Comment (yuyicai) [inline internal/nhctl/controller/dev_start.go:571]: You cat get all `ContainerDevConfig` from `*Controller`.  
No need to add `genContainersAndVolumes()` input parameter `sshUsed bool`

File: internal/nhctl/controller/service.go (hunk 1 above)
Comment (yuyicai) [reply internal/nhctl/controller/service.go:273]: get `language` from `*Controller.config`

File: internal/nhctl/controller/dev_start.go [hunk 5]
```diff
@@ -364,17 +380,25 @@ func (c *Controller) createPvcForPersistentVolumeDir(
...
+			sideCarContainer.SecurityContext = &corev1.SecurityContext{RunAsUser: &rootUID}
+			sideCarContainer.Image = _const.SSHSideCarImage
+		} else {
+			sideCarContainer.Image = _const.DefaultSideCarImage
+		}
+	}
+
 	// over write syncthing command
 	sideCarContainer.Command = []string{"/bin/sh", "-c"}
 	sideCarContainer.Args = []string{
```
Comment (yuyicai) [inline internal/nhctl/controller/dev_start.go:404]: change command

**Comment by yuyicai** [comment]
Change the syncthing-secret volume mode, the default mode is 448 (700). 
The no-root user can not copy the syncthing-secret file (mode 700).

**Comment by yuyicai** [comment]
> Change the syncthing-secret volume mode, the default mode is 448 (700). The no-root user can not copy the syncthing-secret file (mode 700).

deployment test yaml:
``` yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: sec-ctx-demo
  name: sec-ctx-demo
  namespace: default
spec:
  selector:
    matchLabels:
      app: sec-ctx-demo
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        app: sec-ctx-demo
    spec:
      securityContext:
        runAsUser: 1000
      containers:
      - name: sec-ctx-demo
        image: busybox:1.28
        command: [ "sh", "-c", "sleep 1h" ]
```

start dev mode, get error from syncthing sidecar:
```
cp: can't open '/var/syncthing/secret/cert.pem': Permission denied
cp: can't open '/var/syncthing/secret/config.xml': Permission denied
cp: can't open '/var/syncthing/secret/key.pem': Permission denied
```
d3PbX0IFjM

so, we should change the secret mode or change the runAsUser.

**Review by yuyicai — APPROVED** [review]
LGTM  
cc @anurnomeru

**Pull request merged**